# Activity: Explore descriptive statistics

## **Introduction**

Data professionals often use descriptive statistics to understand the data they are working with and provide collaborators with a summary of the relative location of values in the data, as well an information about its spread. 

For this activity, you are a member of an analytics team for the United States Environmental Protection Agency (EPA). You are assigned to analyze data on air quality with respect to carbon monoxide, a major air pollutant. The data includes information from more than 200 sites, identified by state, county, city, and local site names. You will use Python functions to gather statistics about air quality, then share insights with stakeholders.

## **Step 1: Imports** 


Import the relevant Python libraries `pandas` and `numpy`.

In [2]:
# Import relevant Python libraries.

import pandas as pd
import numpy as np



The dataset provided is in the form of a .csv file named `c4_epa_air_quality.csv`. It contains a subset of data from the U.S. EPA. As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [3]:
# RUN THIS CELL TO IMPORT YOUR DATA.

### YOUR CODE HERE
epa_data = pd.read_csv("c4_epa_air_quality.csv", index_col = 0)

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about loading data in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `pandas` library that allows you to read in data from a .csv file and load it into a DataFrame. 

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `read_csv` function from the pandas `library`. The `index_col` parameter can be set to `0` to read in the first column as an index (and to avoid `"Unnamed: 0"` appearing as a column in the resulting DataFrame).

</details>

## **Step 2: Data exploration** 

To understand how the dataset is structured, display the first 10 rows of the data.

In [4]:
# Display first 10 rows of the data.

epa_data.head(10)


date_local    state_name   county_name      city_name  \
0  2018-01-01       Arizona      Maricopa        Buckeye   
1  2018-01-01          Ohio       Belmont      Shadyside   
2  2018-01-01       Wyoming         Teton  Not in a city   
3  2018-01-01  Pennsylvania  Philadelphia   Philadelphia   
4  2018-01-01          Iowa          Polk     Des Moines   
5  2018-01-01        Hawaii      Honolulu  Not in a city   
6  2018-01-01        Hawaii      Honolulu  Not in a city   
7  2018-01-01  Pennsylvania          Erie           Erie   
8  2018-01-01        Hawaii      Honolulu       Honolulu   
9  2018-01-01      Colorado       Larimer   Fort Collins   

                                     local_site_name   parameter_name  \
0                                            BUCKEYE  Carbon monoxide   
1                                          Shadyside  Carbon monoxide   
2  Yellowstone National Park - Old Faithful Snow ...  Carbon monoxide   
3                             North East Waste (NEW)  Carbon monoxide   
4                                          CARPENTER  Carbon monoxide   
5                                            Kapolei  Carbon monoxide   
6                                            Kapolei  Carbon monoxide   
7                                                NaN  Carbon monoxide   
8                                           Honolulu  Carbon monoxide   
9                      Fort Collins - CSU - S. Mason  Carbon monoxide   

    units_of_measure  arithmetic_mean  aqi  
0  Parts per million         0.473684    7  
1  Parts per million         0.263158    5  
2  Parts per million         0.111111    2  
3  Parts per million         0.300000    3  
4  Parts per million         0.215789    3  
5  Parts per million         0.994737   14  
6  Parts per million         0.200000    2  
7  Parts per million         0.200000    2  
8  Parts per million         0.400000    5  
9  Parts per million         0.300000    6

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about exploratory data analysis in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `pandas` library that allows you to get a specific number of rows from the top of a DataFrame. 

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `head()` function from the `pandas` library.

</details>

**Question:** What does the `aqi` column represent?

Aqi stands for air quality index.

Now, get a table that contains some descriptive statistics about the data.

In [6]:
# Get descriptive stats.

epa_data.describe()


arithmetic_mean         aqi
count       260.000000  260.000000
mean          0.403169    6.757692
std           0.317902    7.061707
min           0.000000    0.000000
25%           0.200000    2.000000
50%           0.276315    5.000000
75%           0.516009    9.000000
max           1.921053   50.000000

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `pandas` library that allows you to generate a table of basic descriptive statistics about the numeric columns in a DataFrame.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `describe()` function from the `pandas` library.

</details>

**Question:** Based on the table of descriptive statistics, what do you notice about the count value for the `aqi` column?

The Aqi column has 260 entries just like the arithmetic_mean column.

**Question:** What do you notice about the 25th percentile for the `aqi` column?

This is an important measure for understanding where the aqi values lie. 

The 25th percentile is 2.

**Question:** What do you notice about the 75th percentile for the `aqi` column?

This is another important measure for understanding where the aqi values lie. 

The 75th percentile is 9.

## **Step 3: Statistical tests** 

Next, get some descriptive statistics about the states in the data.

In [8]:
# Get descriptive stats about the states in the data.

desc_stats = epa_data.groupby("state_name").describe()
desc_stats

arithmetic_mean                                          \
                               count      mean       std       min       25%   
state_name                                                                     
Alabama                          1.0  0.200000       NaN  0.200000  0.200000   
Alaska                           2.0  0.555264  0.003722  0.552632  0.553948   
Arizona                         14.0  0.671804  0.519735  0.147368  0.284211   
Arkansas                         1.0  0.300000       NaN  0.300000  0.300000   
California                      66.0  0.684871  0.322950  0.100000  0.420834   
Colorado                         9.0  0.330994  0.071220  0.231579  0.300000   
Connecticut                      4.0  0.200000  0.069537  0.100000  0.187500   
Delaware                         1.0  0.215789       NaN  0.215789  0.215789   
District Of Columbia             2.0  0.222222  0.031427  0.200000  0.211111   
Florida                         12.0  0.299123  0.164393  0.005263  0.231579   
Georgia                          2.0  0.286842  0.122813  0.200000  0.243421   
Hawaii                           3.0  0.531579  0.413384  0.200000  0.300000   
Idaho                            2.0  0.621053  0.379605  0.352632  0.486842   
Illinois                         3.0  0.254386  0.147212  0.142105  0.171052   
Indiana                          5.0  0.304210  0.137820  0.200000  0.231579   
Iowa                             3.0  0.171930  0.062791  0.100000  0.150000   
Kansas                           1.0  0.136842       NaN  0.136842  0.136842   
Kentucky                         3.0  0.181837  0.018426  0.163158  0.172756   
Louisiana                        1.0  0.000000       NaN  0.000000  0.000000   
Maine                            2.0  0.184210  0.022330  0.168421  0.176316   
Maryland                         4.0  0.195833  0.015351  0.183333  0.183992   
Massachusetts                    4.0  0.225000  0.033666  0.184211  0.203947   
Michigan                         9.0  0.465400  0.176400  0.200000  0.378947   
Minnesota                        7.0  0.178947  0.088227  0.010526  0.160527   
Mississippi                      1.0  0.200000       NaN  0.200000  0.200000   
Missouri                         3.0  0.211111  0.019245  0.200000  0.200000   
Montana                          2.0  0.172222  0.102137  0.100000  0.136111   
Nebraska                         2.0  0.324416  0.136666  0.227778  0.276097   
Nevada                           4.0  0.988158  0.603611  0.400000  0.632895   
New Hampshire                    2.0  0.200000  0.000000  0.200000  0.200000   
New Jersey                       6.0  0.302397  0.153683  0.133333  0.187500   
New Mexico                       2.0  0.239474  0.011165  0.231579  0.235526   
New York                        10.0  0.233684  0.041234  0.200000  0.200000   
North Carolina                   4.0  0.191667  0.010638  0.177778  0.186111   
North Dakota                     1.0  0.200000       NaN  0.200000  0.200000   
Ohio                            12.0  0.225379  0.086912  0.083333  0.173143   
Oklahoma                         4.0  0.171930  0.102748  0.036842  0.134211   
Oregon                           2.0  0.152632  0.074432  0.100000  0.126316   
Pennsylvania                    10.0  0.239458  0.122408  0.000000  0.188158   
Puerto Rico                      1.0  0.336842       NaN  0.336842  0.336842   
Rhode Island                     2.0  0.302941  0.145581  0.200000  0.251471   
South Carolina                   1.0  0.100000       NaN  0.100000  0.100000   
South Dakota                     1.0  0.242105       NaN  0.242105  0.242105   
Tennessee                        2.0  0.192105  0.130257  0.100000  0.146053   
Texas                           10.0  0.198947  0.039682  0.131579  0.168421   
Utah                             3.0  0.461306  0.202742  0.326316  0.344737   
Vermont                          3.0  0.227778  0.125093  0.100000  0.166666   
Virginia             

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `pandas` library that allows you to generate basic descriptive statistics about a DataFrame or a column you are interested in.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

 Use the `describe()` function from the `pandas` library. Note that this function can be used:
- "on a DataFrame (to find descriptive statistics about the numeric columns)" 
- "directly on a column containing categorical data (to find pertinent descriptive statistics)"

</details>

**Question:** What do you notice while reviewing the descriptive statistics about the states in the data? 

Note: Sometimes you have to individually calculate statistics. To review to that approach, use the `numpy` library to calculate each of the main statistics in the preceding table for the `aqi` column.

[Write your response here. Double-click (or enter) to edit.]

## **Step 4. Results and evaluation**

Now, compute the mean value from the `aqi` column.

In [9]:
# Compute the mean value from the aqi column.

epa_data["aqi"].mean()

6.757692307692308

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the mean value from an array or a Series of values.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `mean()` function from the `numpy` library.

</details>

**Question:** What do you notice about the mean value from the `aqi` column?

This is an important measure, as it tells you what the average air quality is based on the data.

The mean value is larger than the Q2 which indicates that this dataset is skewed to the left so that there are more larger values than smaller ones or outliers that shift the mean.

Next, compute the median value from the aqi column.

In [10]:
# Compute the median value from the aqi column.
epa_data["aqi"].median()



5.0

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the median value from an array or a series of values.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `median()` function from the `numpy` library.

</details>

**Question:** What do you notice about the median value from the `aqi` column?

This is an important measure for understanding the central location of the data.

The median is below the mean which indicates what has been stated before.

Next, identify the minimum value from the `aqi` column.

In [11]:
# Identify the minimum value from the aqi column.
epa_data["aqi"].min()


0

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the minimum value from an array or a Series of values.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `min()` function from the `numpy` library.

</details>

**Question:** What do you notice about the minimum value from the `aqi` column?

This is an important measure, as it tell you the best air quality observed in the data.

The minimum value is 0 which indicates that 0 is a possible value or that there are outliers that need to be taken care of.

Now, identify the maximum value from the `aqi` column.

In [12]:
# Identify the maximum value from the aqi column.

epa_data["aqi"].max()


50

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the maximum value from an array or a Series of values.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

  Use the `max()` function from the `numpy` library.

</details>

**Question:** What do you notice about the maximum value from the `aqi` column?

This is an important measure, as it tells you which value in the data corresponds to the worst air quality observed in the data.

The max value is 50 which is significantly more than the median or the mean. This further hints that this and maybe other values are skewing the distribution causing the median to differ from the mean.

Now, compute the standard deviation for the `aqi` column.

By default, the `numpy` library uses 0 as the Delta Degrees of Freedom, while `pandas` library uses 1. To get the same value for standard deviation using either library, specify the `ddof` parameter to 1 when calculating standard deviation.

In [13]:
# Compute the standard deviation for the aqi column.

epa_data["aqi"].std()

7.0617066788207215

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

Refer to the video section about descriptive statistics in Python.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the standard deviation from an array or a series of values.

</details>

<details>
  <summary><h4><strong>Hint 3</strong></h4></summary>

Use the `std()` function from the `numpy` library. Make sure to specify the `ddof` parameter as 1. To read more about this function,  refer to its documentation in the references section of this lab.

</details>

**Question:** What do you notice about the standard deviation for the `aqi` column? 

This is an important measure of how spread out the aqi values are.

The standard deviation is of 7.06 which is more than the mean itself, again supporting the idea that outliers are shifting the distribution.

## **Considerations**


**What are some key takeaways that you learned during this lab?**

The data is skewed and outliers are affecting it.

**How would you present your findings from this lab to others? Consider the following relevant points noted by AirNow.gov as you respond:**
- "AQI values at or below 100 are generally thought of as satisfactory. When AQI values are above 100, air quality is considered to be unhealthy—at first for certain sensitive groups of people, then for everyone as AQI values increase."
- "An AQI of 100 for carbon monoxide corresponds to a level of 9.4 parts per million."

All states are within the acceptable range of healthy AQI. However there is quite a disparity between states.

**What summary would you provide to stakeholders? Use the same information provided previously from AirNow.gov as you respond.**

All states are in the healthy threshold however the skew is toward larger values that are distancing from 0.

**References**

[Air Quality Index - A Guide to Air Quality and Your Health](https://www.airnow.gov/sites/default/files/2018-04/aqi_brochure_02_14_0.pdf). (2014,February)

[Numpy.Std — NumPy v1.23 Manual](https://numpy.org/doc/stable/reference/generated/numpy.std.html)

US EPA, OAR. (2014, 8 July).[*Air Data: Air Quality Data Collected at Outdoor Monitors Across the US*](https://www.epa.gov/outdoor-air-quality-data). 

**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.